packages/utils/lib/retry.ts

```diff
@@ -33,6 +35,63 @@ export async function retry<T>(fn: () => T, { maxAttempts, delayMs, retryIf = ()
     throw new Error('unreachable');
 }
 
+export interface RetryAttemptArgument {
+    attempt: number;
+    max: number;
+    waited: number;
+}
+/**
+ * Retry flexible (sic) is a simple re-implementation of exponential backoff that:
+ * - passes the current attempt to the fn
+ * - allow onError to modify the wait time
+ *
+ * It's mostly for our HTTP retry strategy so we can log the attempts and modify the wait time based on headers
+ */
+export async function retryFlexible<TReturn>(
+    fn: (args: RetryAttemptArgument) => TReturn,
+    options: {
+        max: number;
+        /**
+         * Only called if we still have retries available
+         */
+        onError: (arg: { err: unknown; nextWait: number; attempt: number; max: number }) => MaybePromise<{ retry: boolean; reason: string; wait?: number }>;
+    }
+): Promise<TReturn> {
+    const maxDelay = 60 * 15 * 1000; // 15minutes
+    let attempt = -1;
+    let lastWait = 0;
+
+    while (attempt < options.max) {
+        attempt += 1;
+        try {
+            const res = await Promise.resolve(fn({ attempt, max: options.max, waited: lastWait }));
+            return res;
+        } catch (err) {
+            if (attempt >= options.max) {
+                throw err;
+            }
+
+            const nextWait = getExponentialBackoff(attempt, maxDelay);
+            const on = await options.onError({ err, nextWait, max: options.max, attempt: attempt + 1 });
+            if (!on.retry) {
+                throw err;
+            }
+
+            lastWait = on.wait ?? nextWait;
+            await setTimeout(lastWait);
+        }
+    }
+    throw new Error('unreachable');
+}
+
+/**
+ * Get exponential backoff with a cap
+ * Base 2, minimum 1000ms
+ */
+export function getExponentialBackoff(attempt: number, maxDelay: number): number {
+    return Math.min(1000 * 2 ** attempt, maxDelay);
+}
```

Comment:
@bastienbeurier With the new version, I have set up base 1000ms, exp 2, capped to 15minutes. (no cap on maximum retries rn that could break some existing syncs)
Do you want me to change it to something else?

Reply:
This works IMO, but I tweak it to: 
- 3s initial delay
- 2x multiplier for exponential back-off
- 10min max delay
- no max retries

Reply:
Okay I'll change to 3s and 10min. Note we will reach 10 minutes after the 7 retry `0s, 3s, 9s, 21s, 45s, 93s, 189s, 381s, 765s (8th capped), 1533s (9th capped), 3069s (10th capped), etc.`